feat(algebra/lie_algebra): define Lie subalgebras, morphisms, modules, submodules, quotients

ocfnash opened the pull request:
TO CONTRIBUTORS:

Make sure you have:

  * [x] reviewed and applied the coding style: [coding](https://github.com/leanprover/mathlib/blob/master/docs/contribute/style.md), [naming](https://github.com/leanprover/mathlib/blob/master/docs/contribute/naming.md)
  * [x] reviewed and applied [the documentation requirements](https://github.com/leanprover/mathlib/blob/master/docs/contribute/doc.md)
  * [ ] for tactics:
     * [ ] added or adapted documentation in [tactics.md](https://github.com/leanprover/mathlib/blob/master/docs/tactics.md)
     * [ ] write an example of use of the new feature in [tactics.lean](https://github.com/leanprover/mathlib/blob/master/test/tactics.lean)
  * [x] make sure definitions and lemmas are put in the right files
  * [x] make sure definitions and lemmas are not redundant

If this PR is related to a discussion on Zulip, please include a link in the discussion.

For reviewers: [code review check list](https://github.com/leanprover/mathlib/blob/master/docs/contribute/code-review.md)

Trying to knock another chip off https://github.com/leanprover-community/mathlib/issues/1093

Commit:
feat(algebra/lie_algebra): define Lie subalgebras, morphisms, modules, submodules, quotients

jcommelin commented on src/algebra/lie_algebra.lean:
```diff
@@ -174,23 +180,21 @@ end prio
+
+instance (R : Type u) (L : Type v) (L' : Type v)
+  [comm_ring R] [add_comm_group L] [lie_algebra R L] [add_comm_group L'] [lie_algebra R L']
+  : has_coe_to_fun (L →ₗ⁅R⁆ L') := ⟨_, linear_map.to_fun ∘ morphism.to_linear_map⟩
```
Colons go at the end of the previous line.

jcommelin commented on src/algebra/lie_algebra.lean:
```diff
@@ -205,16 +209,191 @@ def of_associative_algebra (A : Type v) [ring A] [algebra R A] : lie_algebra R A
+/--
+The adjoint action of a Lie algebra on itself.
+-/
+def Ad (x : L) : L →ₗ[R] L :=
```
Should this be redefined as a Lie alg morphism from `L` to `module.End R M`?

jcommelin replied:
You do that in `bil_lie_morphism` but that name is quite verbose. I think people wouldn't mind having `Ad` available for the adjoint representation.

ocfnash replied:
I'm not absolutely sure what you have in mind here; would the following capture it:

- rename `Ad` to `Ad_fun`,
- kill `bil_lie` by merging into what is currently called `bil_lie_morphism`,
- rename `bil_lie_morphism` to `Ad`?

jcommelin replied:
Yes, I think that sounds good. (Sorry for the delay, I somehow missed this question.)

ocfnash replied:
I think I've cleaned up the mess here now :-)

jcommelin commented on src/algebra/lie_algebra.lean:
```diff
@@ -205,16 +209,191 @@ def of_associative_algebra (A : Type v) [ring A] [algebra R A] : lie_algebra R A
+A Lie module is a module over a commutative ring, together with a linear action of a Lie algebra
+on this module, such that the Lie bracket acts as the commutator of endomorphisms.
+-/
+class lie_module := (action : L →ₗ⁅R⁆ module.End R M)
```
```suggestion
class lie_module :=
(action : L →ₗ⁅R⁆ module.End R M)
```
I'm not sure if this is the best definition. Usually, in mathlib, such definitions are factored out into "simpler" axioms. See, e.g. `mul_action` and `mul_action_distrib` etc... But this is just some feeling, on my side.
@digama0 Do you have a guideline here?

ocfnash replied:
Yep, I did notice my divergence from the pattern established by `mul_action` etc.

I was back and forth on the approach to take so I'm happy to fall in line with the idiom even if it's just based on a feeling.

I'll wait to hear more here before I mark this as resolved.

ocfnash replied:
Just to check I understand what you have in mind, is something approximately along the lines of the following preferable?
```lean
class add_action (α : Type u) (β : Type v)
  [add_monoid α] [has_zero β] [has_add β] extends has_scalar α β :=
(zero_smul : ∀ (b : β), (0 : α) • b = 0)
(add_smul : ∀ (x y : α) (b : β), (x + y) • b = x • b + y • b)
(smul_add : ∀ (x : α) (b b' : β), x • (b + b') = x • b + x • b')

class linear_action (R : Type u) (L M : Type v)
  [comm_ring R] [add_comm_group L] [add_comm_group M] [module R M] [module R L]
  extends add_action L M :=
(smul_smul  : ∀ (r : R) (l : L) (m : M), (r • l) • m = r • (l • m))
(smul_smul' : ∀ (r : R) (l : L) (m : M), (r • l) • m = l • (r • m))

class lie_module (R : Type u) (L M : Type v)
  [comm_ring R] [add_comm_group L] [add_comm_group M] [module R M] [lie_algebra R L]
  extends linear_action R L M :=
(bracket : ∀ (l l' : L) (m : M), ⁅l, l'⁆ • m = l • (l' • m) - l' • (l • m))
```
(And of course, this could be broken down even further, naming improved, ...)

As I said, I find myself able to argue in favour of either approach. I can imagine that the above might have the convenience of requiring less coercion since the action is definitionally a function, but on the other hand there will presumably be a tonne of lemmas that need to be proved to make these classes convenient.

Perhaps you have other ideas?

All input from @jcommelin and @digama0 gratefully received!

jcommelin replied:
@ocfnash This was indeed what I had in mind (with `bracket` renamed to `lie_smul` :grinning:). I agree that your version is more concise. But it is also less granular. If you have some `linear_action` it can't be nicely promoted into a `lie_module` if it happens to be one. That's one reason why I'm in favour of the granular approach.
Also, concerning the "tonne of lemmas". I fear you'll need them anyway. With either approach, you'll have to write down 15 (or so) simp-lemmas etc, just to get the basics working smoothly.

jcommelin replied:
This is indeed what I had in mind (with `bracket` renamed to `lie_smul` :wink:). The benefit of the granular approach is that objects that already satisfy half of the axioms can gracefully be promoted to a full Lie module. With the 1-line definition, you'll probably make the write a bunch of compatibility glue. (I admit this is mostly intuition, no scientific arguments...)
Concerning the "tonne of lemmas". I guess that with both approaches you'll need to write some 15 (or so) simp-lemmas just to make the basics work smoothly.

ocfnash replied:
Thanks, I've had very little time this week but I'll put this in place over the weekend.

ocfnash replied:
I finally got round to this today, and have just pushed up two commits that aim to define Lie modules in a way closer to what you had in mind.

Annoyingly I was unable to keep the base `add_action` class or to extend `has_scalar` because `linear_action` and `lie_module` both depend on three carrier types and the would-be `add_action` and `has_scalar` depend on only two, which causes problems with the typeclass inference algorithm :-/

The [first commit](https://github.com/leanprover-community/mathlib/pull/1835/commits/0535c44eb284387d50f4175a4990ab5ed53ddd42) shows how I'd like to set things up whereas the [second commit](https://github.com/leanprover-community/mathlib/pull/1835/commits/0f43970204977b025fa1723d314450dc9ff80b96) is what I'm actually proposing.

I tried quite hard to do this well, and also experimented with keeping `add_action` (extending `has_scalar`) but making it external so that you need to supply an `add_action` instance to build a `linear_action` (rather than extending). This does work of course but things seemed to get very messy when I actually tried to use this setup.

Anyway, hopefully getting closer!

jcommelin commented on src/algebra/lie_algebra.lean:
```diff
@@ -205,16 +209,191 @@ def of_associative_algebra (A : Type v) [ring A] [algebra R A] : lie_algebra R A
+This is a sufficient condition for the subset itself to form a Lie algebra.
+-/
+structure lie_subalgebra extends submodule R L :=
+(bracket : ∀ {x y}, x ∈ carrier → y ∈ carrier → ⁅x, y⁆ ∈ carrier)
```
In line with `lie_add`, `lie_neg`, etc... and with `smul_mem` and `add_mem` etc..., this condition should be called `lie_mem` instead of `bracket`.

jcommelin commented on src/algebra/lie_algebra.lean:
```diff
@@ -205,16 +209,191 @@ def of_associative_algebra (A : Type v) [ring A] [algebra R A] : lie_algebra R A
+(bracket : ∀ {x m}, m ∈ carrier → lie_module.action R L M x m ∈ carrier)
+
+instance lie_submodule_coe_submodule [lie_module R L M]
+  : has_coe (lie_submodule R L M) (submodule R M) := ⟨lie_submodule.to_submodule⟩
```
Again, `:` should go on the previous line. Also elsewhere.

jcommelin commented on src/algebra/lie_algebra.lean:
```diff
@@ -205,16 +209,191 @@ def of_associative_algebra (A : Type v) [ring A] [algebra R A] : lie_algebra R A
+  : has_mem M (lie_submodule R L M) := ⟨λ x M', x ∈ (M' : set M)⟩
+
+instance lie_submodule_lie_module [α : lie_module R L M] [M' : lie_submodule R L M]
+  : lie_module R L M' := ⟨{
```
```suggestion
  : lie_module R L M' :=
⟨{
```

jcommelin replied:
Also, this should then be merged with the next line, so that `add` follows the brace.

jcommelin commented on src/algebra/lie_algebra.lean:
```diff
@@ -205,16 +209,191 @@ def of_associative_algebra (A : Type v) [ring A] [algebra R A] : lie_algebra R A
+-/
+abbreviation lie_ideal := lie_submodule R L L
+
+lemma lie_bracket_mem_right (I : lie_ideal R L) (x y : L) (h : y ∈ I) : ⁅x, y⁆ ∈ I := I.bracket h
```
```suggestion
lemma lie_mem_right (I : lie_ideal R L) (x y : L) (h : y ∈ I) : ⁅x, y⁆ ∈ I := I.bracket h
```

jcommelin commented on src/algebra/lie_algebra.lean:
```diff
@@ -205,16 +209,191 @@ def of_associative_algebra (A : Type v) [ring A] [algebra R A] : lie_algebra R A
+
+lemma lie_bracket_mem_right (I : lie_ideal R L) (x y : L) (h : y ∈ I) : ⁅x, y⁆ ∈ I := I.bracket h
+
+lemma lie_bracket_mem_left (I : lie_ideal R L) (x y : L) (h : x ∈ I) : ⁅x, y⁆ ∈ I := by {
```
```suggestion
lemma lie_mem_left (I : lie_ideal R L) (x y : L) (h : x ∈ I) : ⁅x, y⁆ ∈ I := by {
```

jcommelin commented on src/algebra/lie_algebra.lean:
```diff
@@ -205,16 +209,191 @@ def of_associative_algebra (A : Type v) [ring A] [algebra R A] : lie_algebra R A
+lemma lie_bracket_mem_right (I : lie_ideal R L) (x y : L) (h : y ∈ I) : ⁅x, y⁆ ∈ I := I.bracket h
+
+lemma lie_bracket_mem_left (I : lie_ideal R L) (x y : L) (h : x ∈ I) : ⁅x, y⁆ ∈ I := by {
+  rw [←lie_skew, ←neg_lie], apply lie_bracket_mem_right, assumption, }
```
```suggestion
  rw [←lie_skew, ←neg_lie], apply lie_mem_right, assumption, }
```

jcommelin commented on src/algebra/lie_algebra.lean:
```diff
@@ -205,16 +209,191 @@ def of_associative_algebra (A : Type v) [ring A] [algebra R A] : lie_algebra R A
+
+lemma lie_bracket_mem_left (I : lie_ideal R L) (x y : L) (h : x ∈ I) : ⁅x, y⁆ ∈ I := by {
+  rw [←lie_skew, ←neg_lie], apply lie_bracket_mem_right, assumption, }
+
```
Every Lie ideal is a subalgebra. I guess this might be a useful `def` or `instance`.

jcommelin commented on src/algebra/lie_algebra.lean:
```diff
@@ -205,16 +209,191 @@ def of_associative_algebra (A : Type v) [ring A] [algebra R A] : lie_algebra R A
+
+end lie_module
+
+namespace lie_quotient
```
Don't you want to do quotients of Lie modules by submodules in general? If you make the namespace `lie_submodule`, then you can write `N.quotient` for the quotient by the lie submodule `N`.

jcommelin replied:
The downside of not doing the general case is that you would have `N.to_submodule.quotient` all over the place. Just doesn't really look nice.

ocfnash replied:
Agreed, I'll have a bash at quotients by submodules later this evening!

ocfnash replied:
Hmm, I just had a go at doing quotients by submodules. I have it working modulo a bunch of lemmas about `submodule.mapq`, which essentially state that `submodule.mapq` is a linear map, and is functorial, and so gives a morphism of endomorphism algebras when the source and target are the same.

It will take me a little while to put all this together into something sensible. Possible case for separate PR, given that I'll be proposing additions to `linear_algebra/basic.lean`?

jcommelin replied:
If you want to put it in a different PR, that's fine with me? Will you then also remove the bit that currently have on quotients?

ocfnash replied:
Thanks; in that case, I vote for a follow-up PR to handle the quotients (and properties of `submodule.mapq`).

As to whether I should remove what I have on quotients, I will follow your guidance. I tend to think it can stay since I did add the commit to namespace within `lie_submodule` as you suggested, so I think the construction of the quotient is good, and the only thing missing is the following instance:
```lean
instance lie_quotient_lie_module : lie_module R L N.quotient := ...
```

jcommelin replied:
Ok, then let's leave it in. :+1:

jcommelin replied:
Ok, then let's leave it in :+1:

jcommelin commented on src/algebra/lie_algebra.lean:
```diff
@@ -205,16 +209,191 @@ def of_associative_algebra (A : Type v) [ring A] [algebra R A] : lie_algebra R A
-lie_algebra.of_associative_algebra R (matrix n n R)
+lie_algebra.of_associative_algebra (matrix n n R)
+
+#lint
```
This line should be removed at some point.

Commit:
Code review: colons at end of line

Commit:
Update src/algebra/lie_algebra.lean

Co-Authored-By: Johan Commelin <johan@commelin.net>

Commit:
Update src/algebra/lie_algebra.lean

Co-Authored-By: Johan Commelin <johan@commelin.net>

Commit:
Update src/algebra/lie_algebra.lean

Co-Authored-By: Johan Commelin <johan@commelin.net>

Commit:
Catch up after GH commits from code review

Commit:
Remove accidentally-included '#lint'

Commit:
Rename: lie_subalgebra.bracket --> lie_subalgebra.lie_mem

Commit:
Lie ideals are subalgebras

Commit:
Add missing doc string

Commit:
Update src/algebra/lie_algebra.lean

Co-Authored-By: Johan Commelin <johan@commelin.net>

Commit:
Allow quotients of lie_modules by lie_submodules (part 1)

The missing piece is the construction of a lie_module structure on
the quotient by a lie_submodule, i.e.,:
`instance lie_quotient_lie_module : lie_module R L N.quotient := ...`

I will add this in due course.

jcommelin commented on src/algebra/lie_algebra.lean:
```diff
@@ -305,18 +305,18 @@ instance lie_submodule_coe_submodule [lie_module R L M] :
 
-instance lie_submodule_lie_module [α : lie_module R L M] [M' : lie_submodule R L M] :
-  lie_module R L M' :=
+instance lie_submodule_lie_module [α : lie_module R L M] [N : lie_submodule R L M] :
```
Ooh, the `lie_submodule` shouldn't be an instance variable. Use `()` instead of `[]`.

jcommelin commented on src/algebra/lie_algebra.lean:
```diff
@@ -339,24 +339,31 @@ def lie_ideal_subalgebra (I : lie_ideal R L) : lie_subalgebra R L := {
-lemma is_quotient_mk {I : lie_ideal R L} (x : L) :
-  quotient.mk' x = (mk x : quotient I) := by { intros, refl, }
+lemma is_quotient_mk (m : M) :
+  quotient.mk' m = (mk m : N.quotient) := by { intros, refl, }
```
What is `intros` doing here? Can't you replace the entire proof by `rfl`?

ocfnash replied:
Of course! I don't know how I ended up with that 🤦‍♂

jcommelin committed:
Merge branch 'master' into lie_algebra_morphisms_etc

Commit:
Code review: minor fixes

Commit:
Merge branch 'lie_algebra_morphisms_etc' of github.com:ocfnash/mathlib into lie_algebra_morphisms_etc

Commit:
New lie_module approach based on add_action, linear_action

Commit:
Remove add_action by merging into linear_action.

I would prefer to keep add_action, and especially like to keep the feature
that linear_action extends has_scalar, but unfortunately this is not
possible with the current typeclass resolution algorithm since we should
never extend a class with fewer carrier types.

Commit:
Add missing doc string

Commit:
Simplify Lie algebra adjoing action definitions

Commit:
whitespace tweaks

Commit:
Remove redundant explicit type

jcommelin commented on src/algebra/lie_algebra.lean:
```diff
@@ -174,23 +180,26 @@ end prio
+structure morphism (R : Type u) (L : Type v) (L' : Type v)
+  [comm_ring R] [add_comm_group L] [lie_algebra R L] [add_comm_group L'] [lie_algebra R L']
+  extends linear_map R L L' :=
+(bracket : ∀ {x y : L}, to_fun ⁅x, y⁆ = ⁅to_fun x, to_fun y⁆)
```
```suggestion
(map_lie : ∀ {x y : L}, to_fun ⁅x, y⁆ = ⁅to_fun x, to_fun y⁆)
```
Sorry for coming back to these silly issue of names. But I think `map_lie` is more in line with `map_add` and friends. As such, this is the name that users would expect, I guess. Please take a look at the other names, to see if they need updating as well.

ocfnash replied:
It's not silly at all; as a consumer of mathlib I have some appreciation of the value of the good naming that is already in place. I really do try to come up with sensible names, even though that must seem doubtful at times!

jcommelin commented on src/algebra/lie_algebra.lean:
```diff
@@ -174,23 +180,26 @@ end prio
+  [comm_ring R] [add_comm_group L] [lie_algebra R L] [add_comm_group L'] [lie_algebra R L'] :
+  has_coe (L →ₗ⁅R⁆ L') (L →ₗ[R] L') := ⟨morphism.to_linear_map⟩
+
+@[simp] lemma morphism_bracket {R : Type u} {L : Type v} {L' : Type v}
```
```suggestion
@[simp] lemma map_lie {R : Type u} {L : Type v} {L' : Type v}
```

jcommelin commented on src/algebra/lie_algebra.lean:
```diff
@@ -205,16 +214,209 @@ def of_associative_algebra (A : Type v) [ring A] [algebra R A] : lie_algebra R A
-of_associative_algebra R (module.End R M)
+of_associative_algebra (module.End R M)
+
+@[simp] lemma endo_algebra_bracket (M : Type v)
```
```suggestion
lemma endo_algebra_bracket (M : Type v)
```
I'm not sure if this should always be `simp`ed away. I think it's better to let the user add it to the simp set manually. Because once you've simped away the Lie bracket, other simp lemmas involving the bracket will no longer fire.

ocfnash replied:
Very good point, agreed.

jcommelin commented on src/algebra/lie_algebra.lean:
```diff
@@ -205,16 +214,209 @@ def of_associative_algebra (A : Type v) [ring A] [algebra R A] : lie_algebra R A
+    to_fun := has_bracket.bracket x,
+    add    := by { intros, apply lie_add, },
+    smul   := by { intros, apply lie_smul, },
+  },
```
This bracket should go on the previous line.

jcommelin reviewed (commented):
Sorry that it took such a long time to review. I think this is starting to look good.

Commit:
Update src/algebra/lie_algebra.lean

Co-Authored-By: Johan Commelin <johan@commelin.net>

Commit:
Update src/algebra/lie_algebra.lean

Co-Authored-By: Johan Commelin <johan@commelin.net>

Commit:
Update src/algebra/lie_algebra.lean

Co-Authored-By: Johan Commelin <johan@commelin.net>

Commit:
Update src/algebra/lie_algebra.lean

Co-Authored-By: Johan Commelin <johan@commelin.net>

Commit:
Catch up after rename bracket --> map_lie in morphism

jcommelin commented on src/linear_algebra/linear_action.lean:
```diff
@@ -0,0 +1,101 @@
+@[simp] lemma linear_action_add_act [linear_action R M N] (m m' : M) (n : N) :
+  linear_action.act R (m + m') n = linear_action.act R m n +
+                                   linear_action.act R m' n :=
+  linear_action.add_act R m m' n
```
```suggestion
linear_action.add_act R m m' n
```

jcommelin commented on src/linear_algebra/linear_action.lean:
```diff
@@ -0,0 +1,101 @@
+@[simp] lemma linear_action_act_add [linear_action R M N] (m : M) (n n' : N) :
+  linear_action.act R m (n + n') = linear_action.act R m n +
+                                   linear_action.act R m n' :=
+  linear_action.act_add R m n n'
```
```suggestion
linear_action.act_add R m n n'
```

jcommelin commented on src/linear_algebra/linear_action.lean:
```diff
@@ -0,0 +1,101 @@
+
+@[simp] lemma linear_action_act_smul [linear_action R M N] (r : R) (m : M) (n : N) :
+  linear_action.act R (r • m) n = r • (linear_action.act R m n) :=
+  linear_action.act_smul r m n
```
```suggestion
linear_action.act_smul r m n
```

jcommelin commented on src/linear_algebra/linear_action.lean:
```diff
@@ -0,0 +1,101 @@
+
+@[simp] lemma linear_action_smul_act [linear_action R M N] (r : R) (m : M) (n : N) :
+  linear_action.act R m (r • n) = linear_action.act R (r • m) n :=
+  linear_action.smul_act r m n
```
```suggestion
linear_action.smul_act r m n
```

jcommelin commented on src/linear_algebra/linear_action.lean:
```diff
@@ -0,0 +1,101 @@
+/--
+A linear map to the endomorphism algebra yields a linear action.
+-/
+def of_endo_map (α : M →ₗ[R] module.End R N) : linear_action R M N := {
```
```suggestion
def of_endo_map (α : M →ₗ[R] module.End R N) : linear_action R M N :=
```

jcommelin commented on src/linear_algebra/linear_action.lean:
```diff
@@ -0,0 +1,101 @@
+A linear map to the endomorphism algebra yields a linear action.
+-/
+def of_endo_map (α : M →ₗ[R] module.End R N) : linear_action R M N := {
+  act      := λ m n, α m n,
```
```suggestion
{ act      := λ m n, α m n,
```

jcommelin commented on src/linear_algebra/linear_action.lean:
```diff
@@ -0,0 +1,101 @@
+/--
+A linear action yields a linear map to the endomorphism algebra.
+-/
+def to_endo_map (α : linear_action R M N) : M →ₗ[R] module.End R N := {
```
```suggestion
def to_endo_map (α : linear_action R M N) : M →ₗ[R] module.End R N :=
```

jcommelin commented on src/linear_algebra/linear_action.lean:
```diff
@@ -0,0 +1,101 @@
+A linear action yields a linear map to the endomorphism algebra.
+-/
+def to_endo_map (α : linear_action R M N) : M →ₗ[R] module.End R N := {
+  to_fun  := λ m, {
```
```suggestion
{ to_fun  := λ m,
```

jcommelin commented on src/linear_algebra/linear_action.lean:
```diff
@@ -0,0 +1,101 @@
+-/
+def to_endo_map (α : linear_action R M N) : M →ₗ[R] module.End R N := {
+  to_fun  := λ m, {
+    to_fun := λ n, linear_action.act R m n,
```
```suggestion
  { to_fun := λ n, linear_action.act R m n,
```

Commit:
Update src/linear_algebra/linear_action.lean

Co-Authored-By: Johan Commelin <johan@commelin.net>

Commit:
Update src/linear_algebra/linear_action.lean

Co-Authored-By: Johan Commelin <johan@commelin.net>

Commit:
Update src/linear_algebra/linear_action.lean

Co-Authored-By: Johan Commelin <johan@commelin.net>

Commit:
Update src/linear_algebra/linear_action.lean

Co-Authored-By: Johan Commelin <johan@commelin.net>

Commit:
Update src/linear_algebra/linear_action.lean

Co-Authored-By: Johan Commelin <johan@commelin.net>

Commit:
Update src/linear_algebra/linear_action.lean

Co-Authored-By: Johan Commelin <johan@commelin.net>

Commit:
Update src/linear_algebra/linear_action.lean

Co-Authored-By: Johan Commelin <johan@commelin.net>

Commit:
Update src/linear_algebra/linear_action.lean

Co-Authored-By: Johan Commelin <johan@commelin.net>

Commit:
Update src/linear_algebra/linear_action.lean

Co-Authored-By: Johan Commelin <johan@commelin.net>

mergify[bot] committed:
Merge branch 'master' into lie_algebra_morphisms_etc
